lib/link.ts

```diff
@@ -0,0 +1,279 @@
+  }
+
+  /**
+   * Determines whether the remote end of the link is open.
+   * @return {boolean} boolean `true` - is open; `false` otherwise.
+   */
+  isRemoteOpen(): boolean {
+    return this._link.is_remote_open();
+  }
+
+  /**
+   * Determines whether the link has credit.
+   * @return {boolean} boolean `true` - has credit; `false` otherwise.
+   */
+  hasCredit(): boolean {
+    return this._link.has_credit();
+  }
+
+  /**
+   * Determines whether the link is a sender.
+   * @return {boolean} boolean `true` - sender; `false` otherwise.
+   */
+  isSender(): boolean {
+    return this._link.is_sender();
+  }
+
+  /**
+   * Determines whether the link is a receiver.
+   * @return {boolean} boolean `true` - receiver; `false` otherwise.
+   */
+  isReceiver(): boolean {
+    return this._link.is_receiver();
+  }
+
+  /**
+   * Determines whether both local and remote endpoint for link or it's underlying session
+   * or it's underlying connection are closed.
+   * Within the "sender_close", "session_close" event handler, if this
+   * method returns `false` it means that the local end is still open. It can be useful to
+   * determine whether the close was initiated locally under such circumstances.
+   *
+   * @returns {boolean} `true` if closed, `false` otherwise.
+   */
+  isClosed(): boolean {
+    return this._link.is_closed();
+  }
+
+  /**
+   * Determines whether both local and remote endpoint for just the link itself are closed.
+   * Within the "sender_close" event handler, if this method returns `false` it
+   * means that the local end is still open. It can be useful to determine whether the close
+   * was initiated locally under such circumstances.
+   *
+   * @returns {boolean} `true` - closed, `false` otherwise.
+   */
+  isItselfClosed(): boolean {
+    return this._link.is_itself_closed();
+  }
+
+  /**
+   * Determines whether both local and remote endpoint for session or it's underlying
+   * connection are closed.
+   *
+   * Within the "session_close" event handler, if this method returns `false` it means that
+   * the local end is still open. It can be useful to determine whether the close
+   * was initiated locally under such circumstances.
+   *
+   * @returns {boolean} `true` - closed, `false` otherwise.
+   */
+  isSessionClosed(): boolean {
+    return this._session.isClosed();
+  }
+
+  /**
+   * Determines whether both local and remote endpoint for just the session itself are closed.
+   * Within the "session_close" event handler, if this method returns `false` it means that
+   * the local end is still open. It can be useful to determine whether the close
+   * was initiated locally under such circumstances.
+   *
+   * @returns {boolean} `true` - closed, `false` otherwise.
+   */
+  isSessionItselfClosed(): boolean {
+    return this._session.isItselfClosed();
+  }
+
+  /**
+   * Removes the sender and it's underlying session from the internal map.
+   * @returns {void} void
+   */
+  remove(): void {
```

Comment:
this looks equivalent to a dispose. it's a good idea to remove event listeners here, particularly the ones you set up on the underlying link itself. otherwise you can end up with memory leaks because the event emitters are often not cleaned up when there are still active listeners. same goes for sessions and connections